### ISSUE TITLE
"continuous incrementing" slider question

### ISSUE THREAD

**Issue by emily-phet** [phetsims/john-travoltage#267]
@jessegreenberg would it be possible to make John Travoltage arm a custom slider, with the behavior that once the max input value is reached, any typical key press that increment the input value results in jumping the slider back to minimum input value position? If so, this would result in the slider being able to be continuously incremented, without reaching an 'end'.

**Comment by emily-phet** [comment]
Note: I know the current arm slider does not actually span 360 degress, there's a piece of the pie missing(!) opposite the doorknob. I was just curious about this, as I'm writing about the arm/slider implementation for the HCII paper and I'm not sure if we seriously considered this custom approach during our initial design.

**Comment by emily-phet** [comment]
also - I wonder if the initial pause/speed up behavior used in the Coulomb's Law ruler may be something useful for JT.

**Comment by jessegreenberg** [comment]
Regarding
> once the max input value is reached, any typical key press that increment the input value results in jumping the slider back to minimum input value position?

That would be atypical for an HTML slider. Some quick searching found this example of a custom element that wraps: http://roundsliderui.com/demos.html

Regarding:
> I wonder if the initial pause/speed up behavior used in the Coulomb's Law ruler may be something useful for JT.

Sliders get this for free, this is actually in the published version.

I think Jonathan Hung would have some more thoughts about this since he was the original a11y designer for this sim.

**Comment by jessegreenberg** [comment]
> there's a piece of the pie missing(!) opposite the doorknob.

I think this could have been a bug (https://github.com/phetsims/john-travoltage/issues/268), it should wrap all the way around now.

**Comment by emily-phet** [comment]
@jessegreenberg 
Just to clarify - from my initial comments, the only remaining thing is the possibility of a slider that "wraps", correct?
Can you estimate how difficult it would be to implement a wrapping slider? If it's not difficult, it might be useful to try it out with a few students the next time we get access to screen reader users. If it's challenging, probably not worth it.

**Comment by jessegreenberg** [comment]
@emily-phet this would not be a difficult feature to add. With the current design, it would sound something like this:
```
Position negative 10: Away from doorknob.
Position negative 11: Far from doorknob.
Position negative 12: Very far from doorknob.
Position negative 13: Hand pointing away from door, very far from doorknob.
Position negative 14: Very far from doorknob.
Position negative 15: Farthest from doorknob. Last stop.
Position 14: Towards doorknob.
Position 13: Hand pointing away from door, very far from doorknob.
Position 12: Very far from doorknob.
Position 11: Far from doorknob.
```

**Comment by jessegreenberg** [comment]
This would be easy for keyboard input, but may not be possible with gesture input. @emily-phet is it OK if the wrapping feature is supported with keyboard but not gesture?

**Comment by jessegreenberg** [comment]
> presumably though within a gesture context it would behave like a typical slider - correct?

Yes, that is correct.

@mbarlow12 could you please work on this after https://github.com/phetsims/faradays-law/issues/103? since this is a prototype feature like the one in faradays-law this could be worked on in a one-off branch.

**Comment by mbarlow12** [comment]
@jessegreenberg I think I found a slight bug with the AccessibleSlider. The `_constrainValue` function is called on any keydown event if the node is enabled (not just slider inputs). Since the value is immediately set after invocation, this could definitely lead to some odd behavior.

I'll make a new issue, but please correct me if I'm wrong.

**Comment by jessegreenberg** [comment]
Thanks @mbarlow12, I am not sure if this is buggy, but it is inefficient. Is the fix to put everything under line 291 inside of the
```js
if ( KeyboardUtil.isRangeKey( code ) ) {}
```
block?

**Comment by mbarlow12** [comment]
@jessegreenberg That should do it. I'll commit that tonight. Thanks!

**Comment by mbarlow12** [comment]
Implemented in the above commits, and we can review here:

https://www.colorado.edu/physics/phet/dev/html/john-travoltage/1.4.0-circular_slider.1/phet/john-travoltage_en_phet.html?ea&accessibility

**Comment by emily-phet** [comment]
Cool! I put this in the agenda for Tuesday.

**Comment by jessegreenberg** [comment]
We tested in 3/27/18 a11y meeting:

Team really liked the feature. @mbarlow12 were you able to test the changes with VoiceOver? We were curious if this impacted how it sounds with a screen reader. We should also test with NVDA and JAWS.

@terracoda mentioned that we will need to remove the "last stop" phrase during arm movement.  @mbarlow12 could you please do this?

Once it is sounding good we would like to do some user testing.

**Comment by mbarlow12** [comment]
@jessegreenberg my apologies for missing that part of the meeting this AM. I'll hop on the VO testing and the phrasing.

**Comment by mbarlow12** [comment]
There's some odd behavior around the end of the slider range. What I'm noticing is that in VO, the increments will skip the 15th positions when moving across the end of the range. Essentially it move from -15 -> +14 or +15 -> -14. Both end positions can still be reached moving back the other direction.

@emily-phet I assume this is a fairly low priority at the moment?

**Comment by terracoda** [comment]
@mbarlow12, might it have to do that with the fact that Positions -15 and +15 are essentially the same, "Farthest from doorknob. Lat stop."?

If @emily-phet says this is a priority, let me know if you need me to look into what the aria-text values should be as you pass though the far side of the circle.

**Comment by mbarlow12** [comment]
@terracoda while they're the same position visually, they're also on opposite ends of the range input. It may be a bug that arises from flipping the sign on +/-15 to then increment/decrement to +/-14—initial guess at least.

**Comment by jessegreenberg** [comment]
> Essentially it move from -15 -> +14 or +15 -> -14.

Personally, I would also prefer this behavior to it going -15 -> +15 or +15 -> -15.  If that were the case, it would appear that pressing an arrow key at the end of the range would have no effect.

**Comment by jessegreenberg** [comment]
This is no longer low priority as the changes have been merged into master and publication of this sim is coming up with sonification.

**Comment by emily-phet** [comment]
@jessegreenberg @mbarlow12 let's discuss this on Tuesday in dev shorts meeting.

**Comment by emily-phet** [comment]
someone could go ahead and put it on the agenda - i'm about to step into a meeting and will likely forget to do that...

**Comment by jessegreenberg** [comment]
Its on the agenda!

**Comment by emily-phet** [comment]
can someone summarize what happened with this issue? I think a lot of work happened here, but I don't recall ever seeing the requested prototype. 
No more work is needed on this, just wanted to make sure whatever happened here isn't lost in case we want to revisit this in the future.

**Comment by jessegreenberg** [comment]
Here is a link to an initial prototype with this: https://phet-dev.colorado.edu/html/john-travoltage/1.4.0-circular_slider.1/phet/john-travoltage_en_phet.html

Much of the work related to this issue was in converting the initial implementation of accessibility in AppendageNode to use AccessibleSlider.

There is a branch called circularslider that has a commented out block of code with the initial implementation for this, but it is using variables that have since been removed and doesn't work now. But that block and branch will be a good place to start for next time.

**Comment by emily-phet** [comment]
Don't plan on moving forward with this in the foreseeable future.
